src/NativeScript/Calling/FFICall.cpp

```diff
@@ -72,24 +76,101 @@ void FFICall::visitChildren(JSCell* cell, SlotVisitor& visitor) {
     visitor.append(ffiCall->_parameterTypesCells.begin(), ffiCall->_parameterTypesCells.end());
 }
 
-void FFICall::preCall(ExecState* execState, uint8_t* buffer) {
-    if (!this->_argumentCountValidator(this, execState)) {
-        WTF::String message = WTF::String::format("Actual arguments count: \"%lu\". Expected: \"%lu\". ", execState->argumentCount(), this->parametersCount());
-        execState->vm().throwException(execState, createError(execState, message));
-        return;
+CallType FFICall::getCallData(JSCell*, CallData& callData) {
+    callData.native.function = &call;
+    return CallTypeHost;
     }
 
-    // TODO: Check if arguments can be converted
-
-    for (unsigned i = 0; i < execState->argumentCount(); i++) {
-        JSValue argument = execState->uncheckedArgument(i);
-        void* argumentBuffer = reinterpret_cast<void**>(buffer + this->_argsArrayOffset)[i + this->_initialArgumentIndex];
-        JSCell* parameterType = this->_parameterTypesCells[i].get();
-        this->_parameterTypes[i].write(execState, argument, argumentBuffer, parameterType);
+    EncodedJSValue JSC_HOST_CALL FFICall::call(ExecState* execState) {
+        FFICall* callee = jsCast<FFICall*>(execState->callee());
+        Invocation invocation(*callee);
+        ReleasePoolHolder releasePoolHolder;
 
+        callee->preCall(execState, invocation);
+        callee->_invocationHooks.pre(callee, execState, invocation);
         if (execState->hadException()) {
-            return;
+            return JSValue::encode(execState->exception());
+        }
+
+        {
+            JSLock::DropAllLocks locksDropper(execState);
+            ffi_call(callee->_cif, FFI_FN(invocation.function), invocation._buffer + callee->_returnOffset, reinterpret_cast<void**>(invocation._buffer + callee->_argsArrayOffset));
+        }
+
+        JSValue result = callee->_returnType.read(execState, invocation._buffer + callee->_returnOffset, callee->_returnTypeCell.get());
+
+        if (InvocationHook post = callee->_invocationHooks.post) {
+            post(callee, execState, invocation);
         }
+
+        return JSValue::encode(result);
+    }
+
+    JSObject* FFICall::async(ExecState* execState, JSValue thisValue, const ArgList& arguments) {
+        Invocation* invocation = new Invocation(*this);
```

Comment:
I would suggest that you use `__block std::unique_ptr<Invocation>` so that the value could be properly released if we return from line 128 for example.